sparta/sparta/utils/SpartaSharedPointer.hpp

```diff
@@ -68,20 +68,26 @@ namespace sparta
                 }
             }
 
-            uint32_t count;
+            int32_t count;
             PointerT * p = nullptr;
             const bool perform_delete_;
         };
 
         /// Unlink the reference and delete the memory if last to point to it
-        void unlink_() {
-            if(--ref_count_->count == 0) {
-                if(memory_block_) {
-                    memory_block_->alloc->release_(memory_block_);
-                }
-                else {
-                    delete ref_count_;
-                    ref_count_ = 0;
+        void unlink_()
+        {
+            if(SPARTA_EXPECT_FALSE(ref_count_ != nullptr))
```

Comment:
Why don't you assert here as well instead of the branching?

Reply:
I suppose I could move this check to the destructor of SpartaSharedPointer.  That's where this `if` statement means something.  I'll try that.

Reply:
Also, this function is called by the destructor, which shouldn't throw

Reply:
Okay, then you definitely need the check.

Reply:
Actually, I don't.  I can move it safely to the destructor.  All of the functions in SpartaSharedPointer perform this check (in an assert) before calling unlink.  So, good catch.  ;)

Reply:
Nice, this might add a little bit of performance in release mode.